plugins-builtin/src/main/scala/za/co/absa/enceladus/plugins/builtin/errorsender/mq/KafkaErrorSenderPluginImpl.scala

```diff
@@ -0,0 +1,178 @@
+      case Success(params) => params
+      case Failure(e) =>
+        throw new IllegalArgumentException(s"Incompatible parameter map supplied for ${KafkaErrorSenderPluginImpl.getClass.getName}: $params", e)
+    }
+
+    val dfWithErrors = getIndividualErrors(dataFrame, errorSenderParams)
+    val forKafkaDf = prepareDataForKafka(dfWithErrors, errorSenderParams)
+    sendErrorsToKafka(forKafkaDf)
+  }
+
+  /**
+   * Processes the `dataFrame` - errors are exploded (one error = one line) and filtered to conform to the error source (standardization/conformance)
+   * @param dataFrame standardization/conformance output
+   * @param params plugin processing parameters
+   * @return DF with exploded errors and corresponding to the given error source
+   */
+  def getIndividualErrors(dataFrame: DataFrame, params: ErrorSenderPluginParams): DataFrame = {
+    implicit val singleErrorStardardizedEncoder = Encoders.product[SingleErrorStardardized]
+    implicit val dceErrorEncoder = Encoders.product[DceError]
+
+    val allowedErrorCodes = KafkaErrorSenderPluginImpl.errorCodesForSource(params.sourceId)
+
+    val stdErrors = dataFrame
+      .filter(size(col("errCol")) > 0)
+      // only keep columns that are needed for the actual error publishing
+      .select(
+        col(ColumnNames.enceladusRecordId).cast(DataTypes.StringType).as("recordId"),
+        col(ColumnNames.reportDate),
+        explode(col(ColumnNames.errCol)).as("singleError")
+      )
+      .as[SingleErrorStardardized]
+      .map(_.toDceError(params))
+      .filter(entry => allowedErrorCodes.contains(entry.errorCode)) // Std xor Conf error codes
+      .toDF()
+
+    stdErrors
+  }
+
+  def prepareDataForKafka(stdErrors: DataFrame, params: ErrorSenderPluginParams): DataFrame = {
+    log.info(s"Sending errors to kafka topic ${connectionParams.topicName} ...")
+
+    val valueAvroSchemaString = KafkaErrorSenderPlugin.getValueAvroSchemaString
+    val valueSchemaType = KafkaErrorSenderPlugin.getValueStructTypeSchema
+
+    val keyAvroSchemaString = KafkaErrorSenderPlugin.getKeyAvroSchemaString
+
+    val allValueColumns = struct(stdErrors.columns.head, stdErrors.columns.tail: _*)
+    import za.co.absa.abris.avro.functions.to_confluent_avro
+
+    stdErrors.sqlContext.createDataFrame(stdErrors.rdd, valueSchemaType) // forces avsc schema to assure compatible nullability of the DF
+      .select(
+        to_confluent_avro(
+          struct(lit(params.sourceSystem).as("sourceSystem")), keyAvroSchemaString, keySchemaRegistryConfig
+        ).as("key"),
+        to_confluent_avro(allValueColumns, valueAvroSchemaString, valueSchemaRegistryConfig).as("value")
+      )
+  }
+
+  /**
+   * Actual data sending
+   * @param df Dataframe with confluent_avro columns - key & value
+   */
+  private[mq] def sendErrorsToKafka(df: DataFrame): Unit = {
+    require(df.schema.fieldNames.contains("key") && df.schema.fieldNames.contains("value"))
+
+    df.write
```

Comment:
A small late request. Can we add a log with `"Sending $count errors to $kafkaTopic..."` and `"Errors sent to Kafka topic successfully"`?

Also, if there is an error sending data to Kafka we should log it, but carry on, similarly how it is done for sending info files. This is because sending things to Kafka is not critical for job execution.
It might be that this is the current behavior already,

Reply:
Sure, no problem.

Reply:
Done.
Seems to work:
```
20/05/20 15:15:39 INFO mq.KafkaErrorSenderPluginImpl: Sending 430 errors to kafka topic dq.errors.std126 ...

20/05/20 15:15:40 INFO mq.KafkaErrorSenderPluginImpl: 430 errors successfully sent to kafka topic dq.errors.std126
```
There is a new unit integration test checking that the sending is skipped when there are no errors for the particular job type (std/conf).